src/ngx_http_lua_kong_log.c

```diff
@@ -14,45 +14,82 @@
  * limitations under the License.
  */
 
-
 #include "ngx_http_lua_kong_common.h"
 
+#if (NGX_WIN32)
+/*
+** The dynamic log level feature is only supported on UNIX like system,
+** because we only call the function
+** void ngx_http_lua_kong_log_set_ready_for_dynamic_log_level();
+** in ngx_worker_process_cycle in the file src/os/unix/ngx_process_cycle.c.
+**
+** Excluding the technical view,
+** the Kong Gateway does not supoort the Windows OS.
+** 
+*/
+#error "The dynamic log level feature is only supported on UNIX like system"
+#endif
+
+/* -- Global variables for the dynamic log level feature ------------------- */
+
+/* :1 if the dynamic log level is enabled */
+static ngx_uint_t g_dynamic_log_level_set;
+
+/* the dynamic log level */
+static ngx_uint_t g_dynamic_log_level;        
+
+/* after this time, the dynamic log level will restore to original value */
+static time_t     g_dynamic_log_level_timeout_at; 
+
+/* ------------------------------------------------------------------------- */
 
 int
-ngx_http_lua_kong_ffi_set_log_level(ngx_http_request_t *r, int level)
+ngx_http_lua_kong_ffi_set_dynamic_log_level(int log_level, int timeout)
 {
-    ngx_log_t                   *log;
-    ngx_listening_t             *ls;
-    ngx_http_core_loc_conf_t    *clcf;
-    ngx_uint_t                   i;
-
-    if (r == NULL) {
-        return NGX_ERROR;
+    if (timeout == 0) {
+        /*
+        ** if the timeout is 0,
+        ** we will disable the dynamic log level,
+        ** that is, the log_level will be restored to the default setting from log_level in kong.conf.
+        ** and ignore the first parameter.
+        */
+        g_dynamic_log_level_set = 0;
+        return NGX_OK;
     }
 
-    clcf = ngx_http_get_module_loc_conf(r, ngx_http_core_module);
-    clcf->error_log->log_level = (ngx_uint_t) level;
-
-    if (ngx_cycle == NULL) {
+    /*
+    ** if the log level is invalid,
+    ** we will not change the current log level,
+    ** and return an error.
+    */
+    if (log_level > NGX_LOG_DEBUG || log_level < NGX_LOG_STDERR) {
         return NGX_ERROR;
     }
 
-    if (r->connection && r->connection->log) {
-        log = r->connection->log;
+    g_dynamic_log_level = log_level;
+    g_dynamic_log_level_timeout_at = (time_t)ngx_time() + (time_t)timeout;
+    g_dynamic_log_level_set = 1;
 
-    } else {
-        log = ngx_cycle->log;
-    }
+    return NGX_OK;
+}
 
-    /* current request */
-    log->log_level = (ngx_uint_t) level;
+ngx_uint_t
+ngx_http_lua_kong_get_dynamic_log_level(ngx_uint_t current_log_level)
+{
+    if (!g_dynamic_log_level_set) {
+        return current_log_level;
+    }
 
-    /* for each listening socket (new requests) */
-    ls = ngx_cycle->listening.elts;
-    for (i = 0; i < ngx_cycle->listening.nelts; i++) {
-        ls[i].log.log_level = (ngx_uint_t) level;
-        ls[i].logp->log_level = (ngx_uint_t) level;
+    if (g_dynamic_log_level_timeout_at < ngx_time()) {
```

Comment:
I think that we could move this `if (g_dynamic_log_level_timeout_at < ngx_time())` before  `if (!g_dynamic_log_level_set)`

Reply:
I don't understand, could you please explain why?🤔

Reply:
could be:
```
if (g_dynamic_log_level_timeout_at < ngx_time()) {
       g_dynamic_log_level_set = 0;
    }

    if (!g_dynamic_log_level_set) {
        return current_log_level;
    }
```

Reply:
Do these changes bring some benefits?

Here are my thoughts:

In most cases, the value of `g_dynamic_log_level_set` is `0`. If this change is applied, we need to execute one more `if` statement. Please note this function is a hot function.

I believe this change does not cause significant performance degradation, we can apply it. But it does not improve readability for code.